### PR TITLE
Fix Snapshot Abort Not Waiting for Data Nodes

### PR DESCRIPTION
**Pull request by original-brownbear**
This was a really subtle bug that we introduced a long time ago.
If a shard snapshot is in aborted state but hasn't started snapshotting on a node
we can only send the failed notification for it if the shard was actually supposed
to execute on the local node.
Without this fix, if shard snapshots were spread out across at least two data nodes
(so that each data node does not have all the primaries) the abort would actually
never wait on the data nodes. This isn't a big deal with uuid shard generations
but could lead to potential corruption on S3 when using numeric shard generations
(albeit very unlikely now that we have the 3 minute wait there).
Another negative side-effect of this bug was that master would receive a lot more
shard status update messages for aborted shards since each data node not assigned
a primary would send one message for that primary.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -1415,6 +1415,36 @@ public void testRetentionLeasesClearedOnRestore() throws Exception {
...
+        final ActionFuture<CreateSnapshotResponse> snapshotResponse =
+                client().admin().cluster().prepareCreateSnapshot(repoName, snapshotName).setWaitForCompletion(true).execute();
+        waitForBlock(dataNodeName, repoName, TimeValue.timeValueSeconds(30L));
+
+        logger.info("--> abort snapshot");
+        final ActionFuture<AcknowledgedResponse> deleteResponse =
+                client().admin().cluster().prepareDeleteSnapshot(repoName, snapshotName).execute();
+
+        logger.info("--> wait for 5s to give data nodes some time to process the updated shard snapshot status");
+        TimeUnit.SECONDS.sleep(5L);
```
Comment (ywelsch) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:1440]: Can we avoid the sleep here? For example by waiting for the data node to have applied the cluster state with the aborted snapshots, and showing that no outgoing message was sent to the master?

**Review by ywelsch — COMMENTED** [review]
Fix looks good to me. I've left one question on the test. Please double-check that we also properly handle the case where a snapshot is aborted but the node dropped out of the cluster (and that we have tests for that).

**Comment by original-brownbear** [comment]
>  Please double-check that we also properly handle the case where a snapshot is aborted but the node dropped out of the cluster (and that we have tests for that).

Jup this is covered by `org.elasticsearch.snapshots.DedicatedClusterSnapshotRestoreIT#testSnapshotWithStuckNode` as well as the occasional `SnapshotResiliencyTest` that hits this condition also.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java]: Sure thing, sorry for being lazy there => how about https://github.com/elastic/elasticsearch/pull/58214/commits/192508cc964672c7abbf6001c9ecc30da1ecf4d3 ?

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**